x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -141,4 +206,176 @@ public void removeFromCache(final Predicate<CacheKey> predicate) {
+
+        if (event.routingTableChanged()) {
+            final ClusterState clusterState = event.state();
+            final DiscoveryNode localNode = clusterState.getNodes().getLocalNode();
+            assert localNode.isDataNode();
+
+            final boolean shouldSynchronize = hasSearchableSnapshotShards(clusterState, localNode.getId());
+            cacheSyncTask.allowReschedule.set(shouldSynchronize);
+
+            if (shouldSynchronize == false) {
+                logger.trace("canceling cache synchronization task (no searchable snapshots shard(s) assigned to local node)");
+                cacheSyncTask.cancel();
+
+            } else if (cacheSyncTask.isScheduled() == false) {
+                logger.trace("scheduling cache synchronization task (searchable snapshots shard(s) assigned to local node)");
+                cacheSyncTask.rescheduleIfNecessary();
+            }
+        }
+    }
+
+    /**
+     * Synchronize the cache files and dirs.
+     *
+     * This method iterates over all the searchable snapshot shards assigned to the local node in order to execute {@link CacheFile#fsync()}
+     * on cache entries belonging to shards. When at least one {@link CacheFile#fsync()} call returns a non empty set of completed ranges
+     * this method also fsync the shard's snapshot cache directory, which is the parent directory of the cache entries. Note that this
+     * method is best effort as cache entries might be evicted during iterations and cache files/dirs removed from disk.
+     */
+    protected synchronized void synchronizeCache() {
+        if (lifecycleState() != Lifecycle.State.STARTED) {
+            return;
+        }
+        final ClusterState clusterState = clusterService.state();
+        final RoutingNode routingNode = clusterState.getRoutingNodes().node(clusterState.getNodes().getLocalNodeId());
+        assert routingNode != null;
+
+        final long startTimeNanos = threadPool.relativeTimeInNanos();
+        for (ShardRouting shardRouting : routingNode) {
+            if (shardRouting.active()) {
+                final IndexMetadata indexMetadata = clusterState.metadata().getIndexSafe(shardRouting.index());
+                final Settings indexSettings = indexMetadata.getSettings();
+                if (isSearchableSnapshotStore(indexSettings)) {
+                    final ShardId shardId = shardRouting.shardId();
+                    final SnapshotId snapshotId = new SnapshotId(
+                        SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexSettings),
+                        SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSettings)
+                    );
+                    final IndexId indexId = new IndexId(
+                        SNAPSHOT_INDEX_NAME_SETTING.get(indexSettings),
+                        SNAPSHOT_INDEX_ID_SETTING.get(indexSettings)
+                    );
+
+                    boolean syncDirectory = false;
+                    for (Tuple<CacheKey, CacheFile> entry : cache.entries()) {
```

Comment:
This seems like an n^2 algorithm. Since we iterate the cache entries, could we not just fsync the cache files individually as long as they are in the `cache`?

Reply:
++, I would also think there's no need to be tricky here. We already have all the tricky logic for handling a cache file's life-cycle in `CacheFile` => I think I like the idea of just iterating over all `CacheFile` here combined with Henning's other point of having the `CacheFile` register itself for `fsync` in some form. That seems like it wouldn't add new complexities around synchronization and life-cycle beyond what we already have in `CacheFile`?